Show multiple P194 legislatures for a place

henare opened the pull request:
# What does this do?

Shows multiple legislatures next to a place if that place has multiple legislatures returned in the query.

# Why was this needed?

Previously if a place had multiple truthy statements for its Legislative Body For (P194) property then these were not shown (for countries) or resulted in duplicate entries (for FLACS and cities).

# Relevant Issue(s)

Fixes #8.
Fixes #34.

# Implementation notes

# Screenshots
![screenshot from 2018-02-16 13-51-08](https://user-images.githubusercontent.com/48945/36291667-90168b56-1320-11e8-856d-d27dec26e772.png)
![screenshot from 2018-02-16 13-51-30](https://user-images.githubusercontent.com/48945/36291668-905f1b50-1320-11e8-98d9-3d4a2324c8a4.png)
![screenshot from 2018-02-16 13-51-42](https://user-images.githubusercontent.com/48945/36291669-90b3630e-1320-11e8-8bc6-cafa505269ec.png)

# Notes to Reviewer

# Notes to Merger

henare commented on t/query/country_cities.rb:
```diff
@@ -0,0 +1,26 @@
+
+    it 'should only list Los Angeles once' do
+      cities.select { |c| c.name == 'Los Angeles' }.count.must_equal 1
+    end
```
My thinking was that the hierarchy is _'United States of America (Q30)'_  > _'should only list Los Angeles once'_ but I did consider it the way you've described, and even had it that way in one iteration. So that says to me 👍 for this change.

henare commented:
@tmtmtmtm I've merged in the changes from `master` and in doing so addressed the comment about test hierarchy. Can you please review? 👓

tmtmtmtm commented on lib/query/country_divisions.rb:
```diff
@@ -39,5 +39,9 @@ def sparql
     end
+
+    def legislatures(place_id)
+      division_results.select { |r| r[:item] == place_id }.map { |i| i[:legislature] }.uniq
```
Is the `.uniq` at the end of this needed? I can't quite figure out when/why it is, and the tests pass without it, so if there _is_ a scenario where it's important, I think we need a new test.

henare replied:
Added back in with a test in 0b122ee282030ce6c353e9c21f2f956e181e83e0. The example I found was Jerusalem being listed once with 9 of the same legislatures.

tmtmtmtm commented on lib/query/country_divisions.rb:
```diff
@@ -12,8 +12,8 @@ def initialize(id:)
-        DivisionStruct.new(*r.values_at(:item, :population, :legislature, :office, :head))
-      end
+        DivisionStruct.new(*r.values_at(:item, :population, :office, :head), legislatures(r[:item]))
+      end.uniq(&:id)
```
I may be mis-reading this, but isn't this going to throw away data when somewhere has, for example, co-heads of government? (e.g. Northern Ireland)

[We don't see to have pull request previews turned on on this repo, so it's difficult to validate that — @sagepe, is there any reason we can't turn these on here?]

henare replied:
Yes, I think you're right. I included it because I was trying to fix another thing...

>It also ensures only one of each place is displayed on division and city lists. This problem was additionally noted on #8 (comment).

...but that should clearly be part of a separate issue and pull request. A related issue is #22 - fixing the bigger problem behind that will address this. I'll remove ac749ea4495cfb39e5986cb128b39164991851d2, which includes this change, from this pull request.

tmtmtmtm commented on t/query/country_cities.rb:
```diff
@@ -29,4 +29,22 @@
+
+      it 'should only list Los Angeles once' do
+        cities.select { |c| c.name == 'Los Angeles' }.count.must_equal 1
+      end
```
The logic here is still a little odd for my brain. As implemented it looks as if we actively care about the difference in filtering by `id` vs filtering by `name`, but I don't think that's what's really being tested here, is it?

henare replied:
Thanks for pointing this out - it does look a little weird. It's [about to be removed](https://github.com/everypolitician/legislative-explorer/pull/23/files#r168360618) which means it will no longer matter for this PR but I'll keep that in mind for future changes 👍

tmtmtmtm commented:
@heanre I hit submit on that review too soon, sorry! I still think this is heading in roughly the right direction, but I don't think the current implementation (revolving around `uniq`) is quite working. My suspicion, without having had much time to investigate, is that you're probably going to want to throw things through ruby's `group_by` instead, but I haven't really given much thought to how to best collapse the variation in multiple fields, so that might also be a red herring.

henare commented:
@tmtmtmtm As I mentioned [in a comment](https://github.com/everypolitician/legislative-explorer/pull/23#discussion_r168360618), I think the problem was that I was conflating two issues and confusing this pull request 😖 I've opened a [separate issue](https://github.com/everypolitician/legislative-explorer/issues/34) for that and removed [the commit](https://github.com/everypolitician/legislative-explorer/commit/ac749ea4495cfb39e5986cb128b39164991851d2).

✅ I've removed the [superfluous `.uniq` in the `legislatures` methods](https://github.com/everypolitician/legislative-explorer/pull/23#discussion_r168361290).

♻️ Things were getting a little confusing so I've cleanly rebased this on `master`, integrating those changes. Sorry, I know this will make seeing what I've changed since your last review difficult. In this case I thought it was worth it for the overall readability of this PR now 🔮 

It's ready for review again 📋

henare commented on t/query/country_cities.rb:
```diff
@@ -29,4 +29,18 @@
+    describe 'Los Angeles (Q65)' do
+      subject { cities.find { |c| c.id == 'Q65' } }
+
+      it 'should have two legislatures' do
```
It's worth noting that because of the changes in this PR, we do expect there to be two legislatures (i.e. this test is correct and useful). However there is a separate bug, #20, that says there actually should not be two legislatures (i.e. this test is wrong in the context of that bug).

I've included this test because it tests this change in behaviour, ignoring that bug. When that bug is subsequently fixed we can update this test too.

tmtmtmtm commented:
@henare I think your screenshots above are a bit out of date. In particular, the FLACS for Switzerland now look like this:

![screen shot 2018-02-15 at 06 58 45](https://user-images.githubusercontent.com/57483/36244335-bbf826c0-121d-11e8-8da0-082477f246c6.png)

I can't quite follow if this is still a bug in your implementation, or if you now think that this is actually what this PR _should_ be doing (and you're planning to solve the duplication separately). If it's the latter, then I think you need to update more than the screenshots above, but I'm also quite unconvinced that this is an improvement over what's currently happening.

henare commented:
@tmtmtmtm I [had been planning](https://github.com/everypolitician/legislative-explorer/pull/23#issuecomment-365804658) to solve the duplication separately but since that would have meant this wasn't an improvement over what's currently happening I've [added it back](https://github.com/everypolitician/legislative-explorer/pull/23/commits/c8acc35839b57c19b0a786cb2bb169808827b91b). In adding it back I've fixed the implementation that you rightly suspected wasn't quite working.

In doing that I discovered the legislature uniqueness check that was removed is actually needed. I've [added it back](https://github.com/everypolitician/legislative-explorer/pull/23/commits/0b122ee282030ce6c353e9c21f2f956e181e83e0) with a test.

All the screenshots have been updated, as has the PR description.
